src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,245 @@
+            // of this.currentTask since currentTask can be null
+            if (!this.writing.get()) checkAndRethrow();
+
+            // Put new item in queue
+            try {
+                this.queue.put(item);
+            } catch (InterruptedException e) {
+                throw new RuntimeException("Exception while placing item in queue", e);
+            }
+
+            // If queue size is large enough, and not currently writing items, send to executor
+            if (queue.size() >= this.bufferSize && !this.writing.getAndSet(true)) {
+//                checkAndRethrow();
+                this.drain();
+            }
+        }
+
+        /**
+         * Fully drain the writer. This will block till the last attempt to write finishes, then send all remaining
+         * enqueued items to the {@link AsyncWriterPool}'s executor to we written and block till that completes, checking
+         * the result for exceptions.
+         */
+        @Override
+        public void flush() {
+            checkAndRethrow();
+            if (!this.isClosed) {
+                this.drain();
+                checkAndRethrow();
+            }
+        }
+
+        /**
+         * Launch a task in the {@link AsyncWriterPool} to pull items from this writers queue, and write them, till the
+         * queue is empty.
+         */
+        private void drain() {
+            // check the result of the previous task
+            this.currentTask = AsyncWriterPool.this.executor.submit(() -> {
+
+                try {
+                    // BlockingQueue is threadsafe and uses a different lock for both reading and writing
+                    // ends of the queue, so this is safe and avoids an extra copy.
+                    A item = this.queue.poll();
+                    while (item != null) {
+                        this.writer.write(item);
+                        item = this.queue.poll();
+                    }
+                    this.writer.flush();
+                } finally {
+                    this.writing.set(false);
+                }
+                return null;
+            });
+        }
+
+
+        /**
+         * Close the enclosed writer, first blocking and waiting for any ongoing writing to complete, draining any
+         * remaining elements from the queue, checking the result for exceptions / blocking till it is complete, and
+         * finally closing the writer itself.
+         *
+         * @throws IOException if the enclosed writer's close raises an exception
+         */
+        @Override
+        public void close() throws IOException {
+            // NB: We don't need to check on writing here since checkAndRethrow will block till writing is done
+            this.checkAndRethrow();
+            if (!this.isClosed) {
+                this.isClosed = true;
+                this.drain();
+                this.checkAndRethrow(); // Wait for last write to finish
```

Comment:
Again, don't want to block here because this then causes all writers to close() serially which is undesirable.